### ISSUE TITLE
Add custom linear distance decay function for accessibility measures

### ISSUE THREAD

**Issue by shiqin-liu** [healthysustainablecities/global-indicators#111]
The current accessibility method takes the binary approach with hard threshold, which take a rectangular function assuming the value 1 if included in a pre-defined threshold and the value 0 if not (refer [here](https://github.com/global-healthy-liveable-cities/global-indicators/blob/5e2c1133cabd7d740a00e4969cc1ddfd5f961701/process/sp.py#L283) to the code). This method is appropriate for our purpose of indicator aggregation using open data, and passes the suitability analysis in the validation. 

But there is also limitation to this approach, since it assumes that if a destination is slightly farther away from the origin (even by only 2 m: 499m vs 501m) it will have a lower attraction. Thus, there could be improvement in our accessibility measures to allow for custom linear distance decay function. From the literature, I found two methods may be applicable to our framework:

1. Cumulative–Gaussian function: which combines the rectangular function to model the acceptable distance threshold up to a value of perceived closeness, then apply a modified Gaussian function that reflects the influence of distance beyond that acceptable threshold.

<img width="300" alt="Screen Shot 2021-02-18 at 1 53 48 PM" src="https://user-images.githubusercontent.com/46453772/108406800-e3086500-71f0-11eb-84c4-debee1231346.png">

> Reference: Vale, D. S., & Pereira, M. (2017). The influence of the impedance function on gravity-based pedestrian accessibility measures: A comparative analysis. Environment and Planning B: Urban Analytics and City Science, 44(4), 740–763. https://doi.org/https://doi.org/10.1177/0265813516641685

2. Soft threshold approach: which defines a continuous access score for each destination type ranging from 0 to 1
<img width="300" alt="Screen Shot 2021-02-18 at 1 54 18 PM" src="https://user-images.githubusercontent.com/46453772/108406914-08956e80-71f1-11eb-8726-e53b0f76e4c0.png">

> Higgs, C., Badland, H., Simons, K. et al. The Urban Liveability Index: developing a policy-relevant urban liveability composite measure and evaluating associations with transport mode choice. Int J Health Geogr 18, 14 (2019). https://doi.org/10.1186/s12942-019-0178-8)

We discussed the second option before on basecamp, @carlhiggs you may have more thoughts on this since you are the authors of the approach. 

I haven't implement these within our framework yet, but want to share the thought process so far, and keen to hear others suggestion on this.

**Comment by carlhiggs** [comment]
Hi Shirley,

I'll be back off leave from Monday, but just saw your email so thought I'd
give quick jot of thoughts --- in short, I think our parsimonious approach
is justified given our study aims.  I'm writing this from a mobile phone,
sorry if there are predictive text errors!

I think the use of continuous measures is really important in certain
contexts, particularly when accurate measurement of an individual address
is a priority (eg for geocoded survey participants, to evaluate impact of
built environment exposure on some outcome variable, adjusting for
individual sociodemographic covariates).

There were a few reasons we decided not to do this for our study: 1) with
aggregation, hard and soft estimates converge (even at our small area grid
scale, but especially at city scale); 2) it requires measuring some
distance beyond (eg depending on the formula, an access score might
approach 0 around 1600 or more metres; I've previously used 3200m as a
minimum), and in early stage of our study there were concerns about
processing time for both traversal and accounting for network graph buffer;
3) given those reasons (little likely practical difference for the cost to
achieve our aims), the parsimonious approach (500m binary indicators)
seemed justified; 4) policy makers / non quantitative types may find fuzzy
access score hard to interpret (they want clear answers - eg percentage of
people who have access within a policy relevant distance)

To implement a decay like in those images, we'd have to re-measure access
to ensure all addresses were evaluated to some additional minimum distance
eg 1000+ metres.  I don't think it's necessary for this study, although if
another study were to use the method this additional information would be
of value of they were interested in individual not ecological effects.

Having said this, we also calculate a walkability composite measure, and
once indicators are combined in such a measure they less their direct
interpretability and so that advantage of binary indicators is already
lost.  For this reason, in other studies when I have calculated walkability
indices or other composite measures I have used soft threshold scores due
to greater accuracy of reflecting a local context.  Even so, given our
audience of mapping for policy makers and planners, I think the address
complexity is probably not necessary for this study.

Those are my thoughts, and if this was an issue raised by a reviewer that's
how I would respond; I'm happy to discuss further and open to other
viewpoints.

Cheers and enjoy the weekend,

Carl

**Comment by shiqin-liu** [comment]
@carlhiggs Thanks for the quick reply! I totally agree that we should defend ourselves that binary measure is a preferable method for the study (the case study), based on all the reasons you mentioned. And yes, this is one issue raised by the reviewers asking if we can look into more refined methods like a gravity model or distance decay. My initial thought is, if this is easy to do (also suggested by @gboeing  in the comment), we could create a customized function that allows future users to choose which method is most appropriate for their study. This is thinking from an open-source software standpoint. 

So I have been looking up some literature and thinking about which accessibility method would be appropriate to our study and also easy to integrate into our framework within the review timeframe. The Pandana package we used has the accessibility function defined in a way that allows people to retrieve a certain number of nearest points with distance returned, so I am thinking if we could reuse that for the distance decay model (in our current study, we define it in a way that it searches the first nearest point which is appropriate for the binary method). The challenge I have now is as you have mentioned, the later aggregation for the daily living and walkability composite measure. The full_node distance also adds another layer of complexity to this. So any more thoughts on this would be appreciated! 

I will also see if I could put some of my thoughts into codes...

**Comment by carlhiggs** [comment]
Hey Shirley,

I think if we want to implement this for other to use, it shouldn't be too
hard (ie. not an issue for full distance function, composite indicators or
aggregation).

The most straightforward way might be:

1) user must specify minimum traversal distance appropriate to their
maximum accessibility threshold (if these vary by destination).  This will
depend on the formula used (we could even make this user specified, but
supply an optional default eg like in that ULI paper using a sigmoid
transformation).  They also set a setting to indicate usage of continuous
rather than binary indicators.

2) distance evaluation works as usual (including full distance fix) using
the larger evaluation distance

3) accessibility scores are evaluated using the function (default or user
specified), resulting in score from 0 to 1 (instead of binary indicator, 0
or 1)

4) derived indicators are calculated as usual

5) aggregation proceeds as usual (and likely will return a similar result
in many cases).  Technical interpretation will shift: eg. percentage
indicators although approximately similar are no longer strictly
percentages --- although the practical interpretation will be essentially
the same.

I think at this point everything else is equivalent, so the only novel bit
beyond configuration is conditional logic determining which function to use
to evaluate access scores (binary, default continuous, or custom continuous)

I'm sure there are many other ways to do this, but that one should be easy
to implement assuming the custom function is straightforward (eg evaluation
of sample point distance against a fixed threshold, with some decay curve).

What do you think?  (I just realised I haven't seen Geoff's comment yet, he
probably has said similar from what you say)

I'll look at reviewer comments etc when formally back at work from Monday.

Carl

**Comment by shiqin-liu** [comment]
Thanks Carl! I may have overthink some processes. You are correct that changing the method would not affect the aggregation and full distance operation. Below are the three methods I think may be appropriate to use, written in functions, the first is the binary one we used, the second and the third are the two methods I cited at the beginning:

1) Cumulative opportunities (binary)
 1 if d <= access_dist
 0 if d > access_dist
```
def binary_access_score(df, new_name, old_name, access_dist):
    df[new_name] = (df[old_name] <= access_dist).astype("Int64").fillna(0)
    return df[new_name]
```
    
    
2) Soft threshold access score
 1 / (1+ e ^(k *((d-c)/c)))
```
def soft_access_score(df, new_name, old_name, access_dist, k):
    df[new_name] = 1 / (1+numpy.exp(k * ((df[old_name]-access_dist) / access_dist)))
    
    df[new_name] = df[new_name].astype(float).fillna(0)
    return df[new_name]
```

 3) Cumulative-Gaussian
 1 if d <= access_dist ;
 e ^(-1 *((d^2)/k)) if d > access_dist
```
def Cumulative_Gaussian_access_score(df, new_name, old_name, access_dist, k):
    df_access = pd.DataFrame()
    for n, d in zip(new_name, old_name): #loop through each colomn
        df[n] = 0 #create column
        
        df1=df[n].copy()
        df1.loc[df[d] > access_dist] = numpy.exp(-1 * (((df[d]-access_dist)**2) / k))
        df1.loc[df[d] <= access_dist] = 1
        
        df_access = pd.concat([df_access, df1], axis=1, ignore_index=False)

    df_access = df_access.astype(float).fillna(0)
    return df_access
```

Then, the simple way would be to change is [this line of code ](https://github.com/global-healthy-liveable-cities/global-indicators/blob/5e2c1133cabd7d740a00e4969cc1ddfd5f961701/process/sp.py#L283) to the binary function for our purpose. Other users could choose which functions they want to pass for the accessibility scores, with everything else being equivalent. 

Let me know what you think.

**Comment by carlhiggs** [comment]
Hi @shiqin-liu ,

Thanks for taking the lead on this and drafting the functions --- they look good.  

I've just had a review of the pull request changes, and had a few suggestions.  My comments may have been disjointed there so will respond here.

My main two thoughts on amending these are

- we should provide guidance on sensible default values of k for a given access distance threshold (e.g. k=5 is sensible for the soft threshold function, where it seems to me that k=100,000 is required for the cumulative Gaussian function to return similarly equivalent values)
- we don't need to specify the new names within the function; by not worrying about this it should make the function definitions simpler (just need to return a dataframe of values, with names endowed in the implementation code)
- I think the current definition of cumulative function code modifies the input data frame in the df[n]=0 loop (over writes the input data with zeroes)
- we should make sure we amend the variable names and suffixes relating to the word 'binary' in the implementation code; perhaps 'access' is a good generalised alternative prefix?
Geoff also suggested that by doing fillna(0) prior to casting type it should avoid the need of using int64 for a nullable datatype.

Here is an example of refactored functions with the above in mind

```

import pandas as pd
import numpy

def binary_access_score(df, distances, threshold):
    """
    Calculate accessibiity score using binary measure: 1 if access <= threshold, 0 otherwise
    Parameters
    ----------
    df: DataFrame
        DataFrame with origin-destination distances
    distances: list
         list of distances field names
    threshold: int
        access distance threshold
    Returns
    -------
    DataFrame
    """
    scores = (df[distances] <= threshold).fillna(0).astype(int)
    return scores

def soft_access_score(df, distances, threshold=500, k=5):
    """
    Calculate accessibiity score using soft threshold approach:
    1 / (1+ e ^(k *((dist-threshold)/threshold)))
    Parameters
    ----------
    df: DataFrame
        DataFrame with origin-destination distances
    distances: list
        list of field names for distance records
    threshold: int
        access distance threshold
    k: int
        the slope of decay
    Returns
    -------
    DataFrame
    """
    scores = (1 / (1+numpy.exp(k * ((df[distances]-threshold) / threshold)))).fillna(0)
    return scores

def cumulative_gaussian_access_score(df, distances, threshold=500, k=100000):
    """
    Calculate accessibiity score using Cumulative-Gaussian approach:
    1 if d <= threshold ; otherwise, e ^(-1 *((d^2)/k)) if d > threshold
    Parameters
    ----------
    df: DataFrame
        DataFrame with origin-destination distances
    distances: list
        list of field names for distance records
    threshold: int
        access distance threshold
    k: int
        the slope of decay
    Returns
    -------
    DataFrame
    """
    distances = df[distances].copy()
    distances[distances<=threshold] = 1
    distances[distances>threshold] = numpy.exp(-1 * (((distances[distances>threshold]-threshold)**2) / k))
    scores = distances.fillna(0)
    return scores
```

Implementing these, it would look something like
```
#Example dataframe with two columns of distance records
df = pd.DataFrame({'x1':[100,numpy.NaN,300,400,500,600,700,800,900,1000],'x2':[1000,900,800,700,600,500,400,300,200,100]})

# distance record field names
distances = ['x1','x2'] 

# evaluation threshold
threshold = 500

# compare the three functions
for f in [binary_access_score,soft_access_score,cumulative_gaussian_access_score]:
    prefix =  f.__qualname__[:3].lower()
    df[[f'{prefix}_{x}' for x in distances]] = f(df,distances,threshold)        

df
```

And the output looks like

```
       x1    x2  bin_x1  bin_x2    sof_x1    sof_x2    cum_x1    cum_x2
0   100.0  1000       1       0  0.982014  0.006693  1.000000  0.082085
1     NaN   900       0       0  0.000000  0.017986  0.000000  0.201897
2   300.0   800       1       0  0.880797  0.047426  1.000000  0.406570
3   400.0   700       1       0  0.731059  0.119203  1.000000  0.670320
4   500.0   600       1       0  0.500000  0.268941  1.000000  0.904837
5   600.0   500       0       1  0.268941  0.500000  0.904837  1.000000
6   700.0   400       0       1  0.119203  0.731059  0.670320  1.000000
7   800.0   300       0       1  0.047426  0.880797  0.406570  1.000000
8   900.0   200       0       1  0.017986  0.952574  0.201897  1.000000
9  1000.0   100       0       1  0.006693  0.982014  0.082085  1.000000
```

In the script, we imagine that a user has defined their desired function in the set up script, so, if we imagine a variable `threshold_function` is set in the setup script with the name of the desired function:
```
# re-initialise example dataframe
df = pd.DataFrame({'x1':[100,numpy.NaN,300,400,500,600,700,800,900,1000],'x2':[1000,900,800,700,600,500,400,300,200,100]})
distances = ['x1','x2'] 
# define the `threshold_function` variable (presumably accessed via config[`threshold_function`] in the actual code)
threshold_function = binary_access_score
df[[f'access_{x}' for x in distances]] = threshold_function(df,distances,threshold)   
df
```

and the results

```
       x1    x2  access_x1  access_x2
0   100.0  1000          1          0
1     NaN   900          0          0
2   300.0   800          1          0
3   400.0   700          1          0
4   500.0   600          1          0
5   600.0   500          0          1
6   700.0   400          0          1
7   800.0   300          0          1
8   900.0   200          0          1
9  1000.0   100          0          1
```

I guess we'll have to take care with the renamed derived variables that these are modified as required elsewhere, eg. in the aggregation scripts etc.

I wasn't sure how best to contribute these ideas to your pull request, so I'll leave them here for now to see what you and @gboeing think.

I like Geoff's suggestion for using an approach like that used by OSMnx for code standardisation; probably important to share common guidelines given multiple contributors.  Perhaps we can talk about how this might work in the meeting on Friday.

**Comment by shiqin-liu** [comment]
closed by #112
